### ISSUE TITLE
Codegen : output several files instead of one big Tables.scala

### ISSUE THREAD

**Issue by olivergg** [slick/slick#906]
I think it would be great if there was a way to output the generated classes into several files (one per class) instead of having one big Tables.scala file.

**Comment by olivergg** [comment]
@nafg  yep, that was my first idea, but I really don't know how to do that given the current code generation implementation. What about the 3-layers split ? Can you give me an example on the Category+Post tables.
BTW, I also think one could improve the documentation about contributing to the codegen project...writing a Unit test is not that difficult, but it took me a while to get it working.
```
1) clone the slick project somewhere and switch to the branch of https://github.com/slick/slick/pull/987 
2) sbt #run sbt 
3) project testkit #switch to the testkit project
4) testOnly scala.slick.test.codegen.GeneratedCodeTest
5) The generated files are in target/scala-2.XX/src_managed/test/slick-codegen/scala/slick/test/codegen/generated/multiple
```

We can then see that ```PostsTable``` depends on the lazy val ```Categories``` defined in ```CategoriesTable```.

Instead of using ```  self:CGMultipleTables  =>  ```, we could use the proper number of imports for the required tables.

First, one have to make this work on a basic example of the generated code and then (the hard part) would be to generate the code ;)

**Comment by tculshaw** [comment]
Hi,

I've been mucking about with the ofbiz schema in postgresql. It's got a terrifying 833 tables in it, and it's a highly normalised schema.

The Tables.scala generated is over 49,000 lines long. Compiler blows up no matter how many gig I throw at it in SBT_OPTS.

So I'm watching this issue with interest. It's gonna be a terrifying compile anyway even if split up with 833 files and many dependencies via foreign keys.

I'm happy to supply the pg_dump output if you want something hideous to test with. If you can get this working, then you've probably solved most problems anyone is gonna come up against;)

**Comment by olivergg** [comment]
@tculshaw  Are you using the fix from https://github.com/slick/slick/issues/978 ? 
> Compiler blows up no matter how many gig I throw at it in SBT_OPTS.
Have you wait long enough ? Is there any exception ?
Regarding the code generation, one could argue that the generated code could be put in a separate library and then be compiled once to avoid recompilation issues.

**Comment by tculshaw** [comment]
I wait for around 30 minutes and then it blows with an OutOfMemory exception.

Happy to wait an hour and if I can get a good compile run I will certainly jar it up.

My SBT_OPTS are 

export SBT_OPTS="-Xmx2G -XX:+UseConcMarkSweepGC -XX:+CMSClassUnloadingEnabled -XX:PermSize=512M -XX:MaxPermSize=2G -Xss512M"



And the full log...

[info] Compiling 3 Scala sources to /home/tony/workspace/ofbiz-scala/target/scala-2.11/classes...
[error] java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space
[error] Use 'last' for the full log.
> last
[debug] Running task... Cancelable: false, check cycles: false
[debug] 
[debug] Initial source changes: 
[debug]         removed:Set()
[debug]         added: Set(/home/tony/workspace/ofbiz-scala/src/main/scala/com/mentation/ofbiz/Example.scala, /home/tony/workspace/ofbiz-scala/src/main/scala/com/mentation/ofbiz/Tables.scala, /home/tony/workspace/ofbiz-scala/src/main/scala/com/mentation/ofbiz/Domain.scala)
[debug]         modified: Set()
[debug] Removed products: Set()
[debug] Modified external sources: Set()
[debug] Modified binary dependencies: Set()
[debug] Initial directly invalidated sources: Set(/home/tony/workspace/ofbiz-scala/src/main/scala/com/mentation/ofbiz/Example.scala, /home/tony/workspace/ofbiz-scala/src/main/scala/com/mentation/ofbiz/Tables.scala, /home/tony/workspace/ofbiz-scala/src/main/scala/com/mentation/ofbiz/Domain.scala)
[debug] 
[debug] Sources indirectly invalidated by:
[debug]         product: Set()
[debug]         binary dep: Set()
[debug]         external source: Set()
[debug] All initially invalidated sources: Set(/home/tony/workspace/ofbiz-scala/src/main/scala/com/mentation/ofbiz/Example.scala, /home/tony/workspace/ofbiz-scala/src/main/scala/com/mentation/ofbiz/Tables.scala, /home/tony/workspace/ofbiz-scala/src/main/scala/com/mentation/ofbiz/Domain.scala)
[debug] Recompiling all 3 sources: invalidated sources (3) exceeded 50.0% of all sources
[info] Compiling 3 Scala sources to /home/tony/workspace/ofbiz-scala/target/scala-2.11/classes...
[debug] Getting compiler-interface from component compiler for Scala 2.11.4
[debug] Other repositories:
[debug] Default repositories:
[debug]         FileRepository(local,FileConfiguration(true,None),Patterns(ivyPatterns=List(${ivy.home}/local/[organisation]/[module]/(scala_[scalaVersion]/)(sbt_[sbtVersion]/)[revision]/[type]s/[artifact](-[classifier]).[ext]), artifactPatterns=List(${ivy.home}/local/[organisation]/[module]/(scala_[scalaVersion]/)(sbt_[sbtVersion]/)[revision]/[type]s/[artifact](-[classifier]).[ext]), isMavenCompatible=false))
[debug] Getting compiler-interface from component compiler for Scala 2.11.4
[debug] Other repositories:
[debug] Default repositories:
[debug]         FileRepository(local,FileConfiguration(true,None),Patterns(ivyPatterns=List(${ivy.home}/local/[organisation]/[module]/(scala_[scalaVersion]/)(sbt_[sbtVersion]/)[revision]/[type]s/[artifact](-[classifier]).[ext]), artifactPatterns=List(${ivy.home}/local/[organisation]/[module]/(scala_[scalaVersion]/)(sbt_[sbtVersion]/)[revision]/[type]s/[artifact](-[classifier]).[ext]), isMavenCompatible=false))
[debug] Running cached compiler 85ae75c, interfacing (CompilerInterface) with Scala compiler version 2.11.4
[debug] Calling Scala compiler with arguments  (CompilerInterface):
[debug]         -bootclasspath
[debug]         /usr/lib64/jvm/java-1.7.0-openjdk-1.7.0/jre/lib/resources.jar:/usr/lib64/jvm/java-1.7.0-openjdk-1.7.0/jre/lib/rt.jar:/usr/lib64/jvm/java-1.7.0-openjdk-1.7.0/jre/lib/sunrsasign.jar:/usr/lib64/jvm/java-1.7.0-openjdk-1.7.0/jre/lib/jsse.jar:/usr/lib64/jvm/java-1.7.0-openjdk-1.7.0/jre/lib/jce.jar:/usr/lib64/jvm/java-1.7.0-openjdk-1.7.0/jre/lib/charsets.jar:/usr/lib64/jvm/java-1.7.0-openjdk-1.7.0/jre/lib/netx.jar:/usr/lib64/jvm/java-1.7.0-openjdk-1.7.0/jre/lib/plugin.jar:/usr/lib64/jvm/java-1.7.0-openjdk-1.7.0/jre/lib/rhino.jar:/usr/lib64/jvm/java-1.7.0-openjdk-1.7.0/jre/lib/jfr.jar:/usr/lib64/jvm/java-1.7.0-openjdk-1.7.0/jre/classes:/home/tony/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.11.4.jar
[debug]         -classpath
[debug]         /home/tony/workspace/ofbiz-scala/target/scala-2.11/classes:/home/tony/.ivy2/cache/com.typesafe.slick/slick_2.11/bundles/slick_2.11-3.0.0-RC1.jar:/home/tony/.ivy2/cache/org.slf4j/slf4j-api/jars/slf4j-api-1.6.4.jar:/home/tony/.ivy2/cache/com.typesafe/config/bundles/config-1.2.1.jar:/home/tony/.ivy2/cache/org.reactivestreams/reactive-streams/jars/reactive-streams-1.0.0.RC3.jar:/home/tony/.ivy2/cache/com.typesafe.slick/slick-codegen_2.11/jars/slick-codegen_2.11-3.0.0-RC1.jar:/home/tony/.ivy2/cache/org.slf4j/slf4j-nop/jars/slf4j-nop-1.6.4.jar:/home/tony/.ivy2/cache/org.postgresql/postgresql/jars/postgresql-9.3-1103-jdbc41.jar:/home/tony/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.11.4.jar
java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space
        at java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.util.concurrent.FutureTask.get(FutureTask.java:188)
        at sbt.ConcurrentRestrictions$$anon$4.take(ConcurrentRestrictions.scala:196)
        at sbt.Execute.next$1(Execute.scala:88)
        at sbt.Execute.processAll(Execute.scala:91)
        at sbt.Execute.runKeep(Execute.scala:69)
        at sbt.EvaluateTask$.liftedTree1$1(EvaluateTask.scala:183)
        at sbt.EvaluateTask$.sbt$EvaluateTask$$run$1(EvaluateTask.scala:183)
        at sbt.EvaluateTask$.runTask(EvaluateTask.scala:198)
        at sbt.Aggregation$$anonfun$3.apply(Aggregation.scala:67)
        at sbt.Aggregation$$anonfun$3.apply(Aggregation.scala:65)
        at sbt.EvaluateTask$.withStreams(EvaluateTask.scala:153)
        at sbt.Aggregation$.timedRun(Aggregation.scala:65)
        at sbt.Aggregation$.runTasks(Aggregation.scala:74)
        at sbt.Aggregation$$anonfun$applyTasks$1.apply(Aggregation.scala:34)
        at sbt.Aggregation$$anonfun$applyTasks$1.apply(Aggregation.scala:33)
        at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:62)
        at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:62)
        at sbt.Command$.process(Command.scala:95)
        at sbt.MainLoop$$anonfun$1$$anonfun$apply$1.apply(MainLoop.scala:87)
        at sbt.MainLoop$$anonfun$1$$anonfun$apply$1.apply(MainLoop.scala:87)
        at sbt.State$$anon$1.process(State.scala:176)
        at sbt.MainLoop$$anonfun$1.apply(MainLoop.scala:87)
        at sbt.MainLoop$$anonfun$1.apply(MainLoop.scala:87)
        at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:18)
        at sbt.MainLoop$.next(MainLoop.scala:87)
        at sbt.MainLoop$.run(MainLoop.scala:80)
        at sbt.MainLoop$$anonfun$runWithNewLog$1.apply(MainLoop.scala:69)
        at sbt.MainLoop$$anonfun$runWithNewLog$1.apply(MainLoop.scala:66)
        at sbt.Using.apply(Using.scala:25)
        at sbt.MainLoop$.runWithNewLog(MainLoop.scala:66)
        at sbt.MainLoop$.runAndClearLast(MainLoop.scala:49)
        at sbt.MainLoop$.runLoggedLoop(MainLoop.scala:33)
        at sbt.MainLoop$.runLogged(MainLoop.scala:25)
        at sbt.xMain.run(Main.scala:26)
        at xsbt.boot.Launch$$anonfun$run$1.apply(Launch.scala:109)
        at xsbt.boot.Launch$.withContextLoader(Launch.scala:129)
        at xsbt.boot.Launch$.run(Launch.scala:109)
        at xsbt.boot.Launch$$anonfun$apply$1.apply(Launch.scala:36)
        at xsbt.boot.Launch$.launch(Launch.scala:117)
        at xsbt.boot.Launch$.apply(Launch.scala:19)
        at xsbt.boot.Boot$.runImpl(Boot.scala:44)
        at xsbt.boot.Boot$.main(Boot.scala:20)
        at xsbt.boot.Boot.main(Boot.scala)
Caused by: java.lang.OutOfMemoryError: Java heap space
        at java.util.jar.Manifest$FastInputStream.<init>(Manifest.java:332)
        at java.util.jar.Manifest$FastInputStream.<init>(Manifest.java:327)
        at java.util.jar.Manifest.read(Manifest.java:195)
        at java.util.jar.Manifest.<init>(Manifest.java:69)
        at java.util.jar.JarFile.getManifestFromReference(JarFile.java:185)
        at java.util.jar.JarFile.getManifest(JarFile.java:166)
        at sun.misc.URLClassPath$JarLoader$2.getManifest(URLClassPath.java:779)
        at java.net.URLClassLoader.defineClass(URLClassLoader.java:416)
        at java.net.URLClassLoader.access$100(URLClassLoader.java:71)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
        at scala.tools.nsc.typechecker.Namers$Namer.typeErrorHandler(Namers.scala:111)
        at scala.tools.nsc.typechecker.Namers$Namer.typeSig(Namers.scala:1539)
        at scala.tools.nsc.typechecker.Namers$Namer$$anonfun$monoTypeCompleter$1$$anonfun$apply$1.apply$mcV$sp(Namers.scala:778)
        at scala.tools.nsc.typechecker.Namers$Namer$$anonfun$monoTypeCompleter$1$$anonfun$apply$1.apply(Namers.scala:777)
        at scala.tools.nsc.typechecker.Namers$Namer$$anonfun$monoTypeCompleter$1$$anonfun$apply$1.apply(Namers.scala:777)
        at scala.tools.nsc.typechecker.Namers$Namer.scala$tools$nsc$typechecker$Namers$Namer$$logAndValidate(Namers.scala:1565)
        at scala.tools.nsc.typechecker.Namers$Namer$$anonfun$monoTypeCompleter$1.apply(Namers.scala:777)
        at scala.tools.nsc.typechecker.Namers$Namer$$anonfun$monoTypeCompleter$1.apply(Namers.scala:769)
        at scala.tools.nsc.typechecker.Namers$$anon$1.completeImpl(Namers.scala:1681)
        at scala.tools.nsc.typechecker.Namers$LockingTypeCompleter$class.complete(Namers.scala:1689)
        at scala.tools.nsc.typechecker.Namers$$anon$1.complete(Namers.scala:1679)
        at scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1481)
        at scala.reflect.internal.Symbols$Symbol.initialize(Symbols.scala:1628)
        at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:4911)
        at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5295)
        at scala.tools.nsc.typechecker.Typers$Typer.scala$tools$nsc$typechecker$Typers$Typer$$typedInternal(Typers.scala:5322)
        at scala.tools.nsc.typechecker.Typers$Typer.body$2(Typers.scala:5269)
[error] java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space

**Comment by olivergg** [comment]
@tculshaw  Have you tried with Java 8 ?

**Comment by tculshaw** [comment]
Good suggestion, but similar result:( 

Java 8 gives...
java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: GC overhead limit exceeded

with .bashrc setting...
export SBT_OPTS="-Xmx2G -XX:+UseConcMarkSweepGC -XX:+CMSClassUnloadingEnabled -XX:PermSize=512M -XX:MaxPermSize=2G -Xss512M"

Detail being...
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
        at scala.reflect.internal.Types$Type.substSym(Types.scala:715)
        at scala.reflect.internal.Symbols$class.createFromClonedSymbolsAtOwner(Symbols.scala:3624)
        at scala.reflect.internal.SymbolTable.createFromClonedSymbolsAtOwner(SymbolTable.scala:16)
        at scala.reflect.internal.Types$ExistentialType.cloneInfo(Types.scala:2682)
        at scala.reflect.internal.Symbols$Symbol.cloneSymbol(Symbols.scala:1935)
        at scala.reflect.internal.Symbols$Symbol.cloneSymbol(Symbols.scala:1930)
        at scala.reflect.internal.Symbols$Symbol.cloneSymbol(Symbols.scala:1928)
        at scala.reflect.internal.Symbols$Symbol.cloneSymbol(Symbols.scala:1924)
        at scala.reflect.internal.Symbols$$anonfun$cloneSymbols$1.apply(Symbols.scala:3600)
        at scala.reflect.internal.Symbols$$anonfun$cloneSymbols$1.apply(Symbols.scala:3600)
        at scala.reflect.internal.util.Collections$class.mapList(Collections.scala:56)
        at scala.reflect.internal.SymbolTable.mapList(SymbolTable.scala:16)
        at scala.reflect.internal.Symbols$class.deriveSymbols(Symbols.scala:3539)
        at scala.reflect.internal.SymbolTable.deriveSymbols(SymbolTable.scala:16)
        at scala.reflect.internal.Symbols$class.cloneSymbols(Symbols.scala:3600)
        at scala.reflect.internal.SymbolTable.cloneSymbols(SymbolTable.scala:16)
        at scala.reflect.internal.Symbols$class.cloneSymbolsAndModify(Symbols.scala:3611)
        at scala.reflect.internal.SymbolTable.cloneSymbolsAndModify(SymbolTable.scala:16)
        at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:245)
        at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:128)
        at scala.reflect.internal.tpe.TypeMaps$SubstMap.apply(TypeMaps.scala:700)
        at scala.reflect.internal.Types$Type.subst(Types.scala:705)
        at scala.reflect.internal.Types$Type.instantiateTypeParams(Types.scala:470)
        at scala.reflect.internal.Symbols$class.deriveTypeWithWildcards(Symbols.scala:3595)
        at scala.reflect.internal.SymbolTable.deriveTypeWithWildcards(SymbolTable.scala:16)
        at scala.tools.nsc.typechecker.Implicits$class.scala$tools$nsc$typechecker$Implicits$$depoly(Implicits.scala:154)
        at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.scala$tools$nsc$typechecker$Implicits$ImplicitSearch$$matchesPt(Implicits.scala:484)
        at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation.survives(Implicits.scala:818)
        at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation$$anonfun$19$$anonfun$20.apply(Implicits.scala:872)
        at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation$$anonfun$19$$anonfun$20.apply(Implicits.scala:872)
        at scala.collection.TraversableLike$$anonfun$filterImpl$1.apply(TraversableLike.scala:259)
        at scala.collection.immutable.List.foreach(List.scala:381)

**Comment by tculshaw** [comment]
Another thought, the primarykey types could be causing the codegen problems? Some primary keys are string based, some primary keys are composite primary keys.

**Comment by olivergg** [comment]
@tculshaw  How much physical memory do you have ? Could you try to increase the max heap size ? (-Xmx) ? If you use java 8, you can also drop the "-XX:PermSize=512M -XX:MaxPermSize=2G" flags.

**Comment by virusdave** [comment]
Any progress on this issue?

**Comment by tculshaw** [comment]
I haven't pursued this - sorry!

I'm using slick successfully on a number of smaller schemas though.

Tony Culshaw MA Cantab
Director
Mentation Limited - Custom Software Development

**Comment by cvogt** [comment]
If there are any students willing to work on this, we can offer to do this (and more) through the Google Summer of Code program, which funds students with 5K to work on FOSS projects over the summer.

**Comment by sumiet** [comment]
Any updates on this issue?

**Comment by pettyjamesm** [comment]
Would an improvement here be the ability to specify which schemas, tables, and columns are a part of the generated code?

I appreciate that at some point the internals of the scala compiler become the limiting factor, but I suppose I'm wondering how often this many tables are actually used by logic written in slick.

**Comment by virusdave** [comment]
Yes it's technically an improvement, but it's a rather limited one.  I did
this last january as a mitigation.  My experience was that the number of
tables i needed generated grew slowly but steadily over time, to the point
where 6 months later compiling the single file has once again become a
major factor in compilation time.  Today it's excruciatingly slow.

I doubt i'm alone in this.

  -dave-

**Comment by hvesalai** [comment]
Solved by #1785
